config/secret.go

```diff
@@ -0,0 +1,208 @@
+type Secret struct {
+	enclave   *memguard.Enclave
+	resolvers map[string]telegraf.ResolveFunc
+	// unlinked contains all references in the secret that are not yet
+	// linked to the corresponding secret store.
+	unlinked []string
+
+	// Denotes if the secret is completely empty
+	empty bool
+}
+
+// NewSecret creates a new secret from the given bytes
+func NewSecret(b []byte) Secret {
+	s := Secret{}
+	s.init(b)
+	return s
+}
+
+// UnmarshalTOML creates a secret from a toml value.
+func (s *Secret) UnmarshalTOML(b []byte) error {
+	// Unmarshal raw secret from TOML and put it into protected memory
+	s.init(b)
+
+	// Keep track of secrets that contain references to secret-stores
+	// for later resolving by the config.
+	if len(s.unlinked) > 0 && !s.empty {
+		unlinkedSecrets = append(unlinkedSecrets, s)
+	}
+
+	return nil
+}
+
+// Initialize the secret content
+func (s *Secret) init(b []byte) {
+	secret := unquoteTomlString(b)
+
+	// Remember if the secret is completely empty
+	s.empty = len(secret) == 0
+
+	// Find all parts that need to be resolved and return them
+	s.unlinked = secretPattern.FindAllString(string(secret), -1)
+
+	// Setup the enclave
+	s.enclave = memguard.NewEnclave(secret)
+	s.resolvers = nil
+}
+
+// Destroy the secret content
+func (s *Secret) Destroy() {
+	s.resolvers = nil
+	s.unlinked = nil
+	s.empty = true
+
+	if s.enclave == nil {
+		return
+	}
+
+	// Wipe the secret from memory
+	lockbuf, err := s.enclave.Open()
+	if err == nil {
+		lockbuf.Destroy()
+	}
+	s.enclave = nil
+}
+
+// Empty return if the secret is completely empty
+func (s *Secret) Empty() bool {
+	return s.empty
+}
+
+// Get return the string representation of the secret
+func (s *Secret) Get() ([]byte, error) {
+	if s.enclave == nil {
+		return nil, nil
+	}
+
+	if len(s.unlinked) > 0 {
+		return nil, fmt.Errorf("unlinked parts in secret: %v", strings.Join(s.unlinked, ";"))
+	}
+
+	// Decrypt the secret so we can return it
+	lockbuf, err := s.enclave.Open()
+	if err != nil {
+		return nil, fmt.Errorf("opening enclave failed: %v", err)
+	}
+	defer lockbuf.Destroy()
+	secret := lockbuf.Bytes()
+
+	if len(s.resolvers) == 0 {
+		// Make a copy as we cannot access lockbuf after Destroy, i.e.
+		// after this function finishes.
+		newsecret := append([]byte{}, secret...)
+		return newsecret, protect(newsecret)
+	}
+
+	replaceErrs := make([]string, 0)
+	newsecret := secretPattern.ReplaceAllFunc(secret, func(match []byte) []byte {
+		resolver, found := s.resolvers[string(match)]
+		if !found {
+			replaceErrs = append(replaceErrs, fmt.Sprintf("no resolver for %q", match))
+			return match
+		}
+		replacement, _, err := resolver()
+		if err != nil {
+			replaceErrs = append(replaceErrs, fmt.Sprintf("resolving %q failed: %v", match, err))
+			return match
+		}
+
+		return replacement
+	})
+	if len(replaceErrs) > 0 {
+		return nil, fmt.Errorf("replacing secrets failed: %s", strings.Join(replaceErrs, ";"))
+	}
```

Comment:
I didn't think about this before, but is it possible that `newsecret` needs to be wiped when `len(replaceErrs) > 0`? If so, please do so in the original PR, this PR or in a followup.

Reply:
It cannot hurt to wipe it... Done.